Corner point "click in" editing mode

lpechacek opened the pull request:
Implementation of GH-1345 enhancement.
![out-1_hMSkB5v9_knQo ogv](https://user-images.githubusercontent.com/5584406/64612793-f2b53480-d3d4-11e9-8831-bb8681a62264.gif)

dg0yt commented on src/core/objects/object_mover.cpp:
```diff
@@ -33,7 +33,11 @@
+ , corner_tolerance(0)
+ , prev_drag_x(0)
+ , prev_drag_y(0)
+ , constraints_calculated(true)
```
I would prefer to move the constant initializations to the member variable declaration.

lpechacek replied:
Me too. Assuming that you are talking about all the data members, I'll rather move the refactoring to a separate patch so that I don't mix unrelated changes into the feature implementation.

dg0yt commented on src/core/objects/object_mover.cpp:
```diff
@@ -102,7 +111,30 @@ void ObjectMover::move(const MapCoordF& cursor_pos, bool move_opposite_handles,
-void ObjectMover::move(qint32 dx, qint32 dy, bool move_opposite_handles)
+namespace {
+bool forms_a_corner(const MapCoord anchor_point, const MapCoord moved_handle,
+                    const MapCoord opposite_handle, const qreal quantum_size)
```
`forms_a_corner` does not conform to our spelling of function names.

I'm not so happy with having this function between the ObjectMover member function definitions, but I wonder what is the best place/namespace to put this function to. It might be a candidate for a static function in MapCoord, or at least a free function in map_coord.h. I realized that functions in the anonymous namespace are unaccessible to unit tests, too...

lpechacek replied:
Ack. I originally thought I'll reuse the function in OCD export code for corner point generation but later my thoughts went in different direction. I'll take another at the function naming and placement.

lpechacek replied:
This is now `Util::pointsFormCorner()`.

dg0yt commented on src/core/objects/object_mover.cpp:
```diff
@@ -124,19 +156,35 @@ void ObjectMover::move(qint32 dx, qint32 dy, bool move_opposite_handles)
+
+		// Check conditions for "corner point click-in" operation
+		if (move_opposite_handles == HandleOpMode::Click
+			&& constraint.anchor_is_corner)
```
Please align `&&` with the `move_...` in the line above.

lpechacek replied:
Well, they are aligned with 4-space-sized tab. :)

However, looking again at this part I saw that the mode handling is somewhat confusing in its current form so I rewrote this part into a switch with fallthroughs, which is structuring the code in a more readable manner.

dg0yt commented on src/core/objects/object_mover.cpp:
```diff
@@ -124,19 +156,35 @@ void ObjectMover::move(qint32 dx, qint32 dy, bool move_opposite_handles)
-			constraint.object->setCoordinate(constraint.opposite_handle_index, control);
+			if (forms_a_corner(anchor_point, moved_handle, control, corner_tolerance))
+				break; // still is a corner point -> refrain from moving opposite handle
+			else
```
`else` after `break` is redundant.

lpechacek replied:
Ack.

lpechacek replied:
This part has been reimplemented as a `switch`.

dg0yt commented on src/core/objects/object_mover.h:
```diff
@@ -52,6 +52,15 @@ class ObjectMover
+	 * @param corner_tolerance Maximum difference in vector directions
+	 *                         in millimeters.
+	 */
+	void setCornerTolerance(const qreal corner_tolerance);
```
`const` is fine in the definition, but has no effect and shall be omitted in the declaration.

lpechacek replied:
Ack.

dg0yt commented on src/core/objects/object_mover.h:
```diff
@@ -64,16 +73,24 @@ class ObjectMover
+		Click,
+		Always
+	};
+
```
I would prefer to have the enum at the beginning of the class definition.

dg0yt replied:
I think that this enum and its members need documentation.

lpechacek replied:
Indeed. The `enum` placement is somewhat random. Will fix.

lpechacek commented:
Thanks for the review comments, Kai. I'll post updated patch set ASAP.

dg0yt commented:
:clock1130: Kindly pinging ;-)

lpechacek commented:
😀 yes, this pull request is not forgotten. It was just a conference, sickness and next SUSE Linux release planning which got into the way. Apologies for the slow progress and thanks for the ping.

lpechacek commented:
Changes in this iteration:
* Rebased on top of a newer `master`.
* The `if-else` maze was rewritten as a switch.
* Following [YAGNI principle](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it), `HandleOpMode::Always` was dropped.
* `HandleOpMode` values have been documented.
* Corner point test function has been renamed, moved to `Util` class and documented.
* All refactoring changes are separated from the function change.

Items for discussion:
* Point coordinates are passed to `Util::pointsFormCorner()` by reference, following `isPointOnSegment()` example. While this approach saves compile time, it perhaps induces pointer access in run time. Wouldn't pass-by-value be more efficient at the expense of including `map_coord.h` in `util.h`?

lpechacek commented:
> Point coordinates are passed to Util::pointsFormCorner() by reference, ...

So I looked at function assembly code and compared it with the case when coordinates are passed by value. Without analyzing the very details, I would say the machine code is very similar in both cases on x86_64 and I don't expect significant impact from the reference passing. Of course, the ultimate answer can be provided only through performance measurements but we are not designing a latency critical inner loop here, are we? That said, I consider this point a non-issue from performance point of view and I'm fine if it is ignored.

dg0yt commented:
> > Point coordinates are passed to Util::pointsFormCorner() by reference, ...
> 
> So I looked at function assembly code and compared it with the case when coordinates are passed by value. Without analyzing the very details, I would say the machine code is very similar in both cases on x86_64 and I don't expect significant impact from the reference passing. Of course, the ultimate answer can be provided only through performance measurements but we are not designing a latency critical inner loop here, are we? That said, I consider this point a non-issue from performance point of view and I'm fine if it is ignored.

Qt passes QPoint by const reference, e.g. [https://doc.qt.io/qt-5/qpainter.html#drawEllipse-4](in QPainter::drawEllipse], and I made Mapper follow this standard for MapCoord and friends a year ago, https://github.com/OpenOrienteering/mapper/compare/35526b9...37bedd84

dg0yt commented on src/util/util.h:
```diff
@@ -201,6 +202,25 @@ void gridOperation(const QRectF& extent, double horz_spacing, double vert_spacin
+ *         anchor_point as point1.
+ */
+bool pointsFormCorner(const MapCoord& anchor_point, const MapCoord& point1,
+                      const MapCoord& point2, const qreal quantum_size);
```
- There is the `const qreal` in the declaration again ([readability-avoid-const-params-in-decls](https://releases.llvm.org/8.0.0/tools/clang/tools/extra/docs/clang-tidy/checks/readability-avoid-const-params-in-decls.html)).
- I would appreciate a short unit test for this function at least, in `util_t.cpp`.
- I'm concerned that the MapCoord parameter order is not so obvious when looking at a calling line. How about this signature:
~~~
bool pointsFormCorner(const MapCoord& first_point, const MapCoord& inner_point,
                      const MapCoord& last_point, const qreal quantum_size);
~~~
where `inner_point` takes the role of the former `anchor_point`?

lpechacek replied:
Thanks, Kai, for the quick review. I can see you started using clang-tidy. I looked into making the clang helpers work with openSUSE yesterday but it looks like there is something broken in the tidy plugin handling. Will try harder to make it work.
I'm all fine with `const` removal from the declaration. I'll watch for this error next time. Regarding the parameter ordering for `Util::pointsFormCorner()`, the rename brought a very slight shift in the meaning which escaped my attention. With the current function name it makes sense to order them as "first, middle, last". The original meaning was "(test that a point) forms a corner" and the order was "point to be tested, the other two points".
Ad the test for `Util::pointsFormCorner()`, I'll add a patch to the series.

dg0yt replied:
I was regularly using iwyu and clang-tidy, for the files white listed in code-check-wrapper. But unfortunately I didn't realize that the '.clang-tidy' was ignored due to wrong syntax.

lpechacek replied:
So let's make a move on this one. I've incorporated the drop of `const` on function prototype and parameter reordering for `Util::pointsFormCorner()`.
I've also taken another look at `code-check-wrapper.sh` and maybe I've guessed how to use it. In any case, I'd be happy to see instructions for use in developer wiki if the check output is important part of review.

dg0yt commented on src/core/objects/object_mover.cpp:
```diff
@@ -124,18 +131,31 @@ void ObjectMover::move(qint32 dx, qint32 dy, bool move_opposite_handles)
+		MapCoord anchor_point = constraint.object->getCoordinate(constraint.curve_anchor_index);
+		MapCoord moved_handle = constraint.object->getCoordinate(constraint.moved_handle_index);
+		MapCoordF to_hover_point = MapCoordF(moved_handle - anchor_point);
+		MapCoord control = constraint.object->getCoordinate(constraint.opposite_handle_index);
```
Now looking at this again, I wonder if we can't use a better name than `control` here, e.g. `opposite_handle`.

lpechacek replied:
Well, from my point of view the rename is out of scope of this change. The task is to implement corner click-in editing mode and this is the original code that gets rearranged towards the desired goal. By default I don't change the original code more than necessary. That said, I don't have strong opinion about the naming and I'm fine with either leaving it as is, which is expressed by this patch set, or renaming it to obviously more logical `opposite_handle`.

dg0yt replied:
I know that the name comes from the original code. I just can't skip taking such a note when the unit is touched anyway. It can go to a separate commit, but I wouldn't want to enforce a separate PR, with all the CI cycles.

lpechacek replied:
Fine with me and I don't object incremental refactoring. IMO all these changes can easily go into the last patch, now labeled "Use in-class member initializers ".

Talking about these exact numbers, why stop at the rename? Mapper Coding conventions encourage use of AAA. Guidelines add ES.23 and Con.4. Opinions?
```suggestion
		const auto anchor_point = constraint.object->getCoordinate(constraint.curve_anchor_index);
		const auto moved_handle = constraint.object->getCoordinate(constraint.moved_handle_index);
		auto opposite_handle = constraint.object->getCoordinate(constraint.opposite_handle_index);
		auto to_hover_point = MapCoordF { moved_handle - anchor_point };
```

lpechacek replied:
Accepting the suggestion would break the code as it renames only the variable declaration but not the subsequent use. :)
Thanks for clarification on the expected extent of refactoring. For now I've just made the rename as part of the feature implementation.

lpechacek commented:
> Let me know if I you want to leave this to me.

I don't object to maintainer edits and I believe they will serve the purpose of smooth cooperation. After the years I'm still unable to get all the coding details right and related pull request conversations tend to be lengthy. Feel free to implement small adjustments on your own. I'd appreciate if the changes were mentioned in the commit message like in [this example](https://kernel.suse.com/cgit/kernel-livepatch/commit/?h=SLE15_Update_15&id=f842fd53a604965a864d3827ab90c691b44cb9a6) (note the remarks in square brackets) but it is not an absolute requirement from my side.

dg0yt commented:
> > Let me know if I you want to leave this to me.
> 
> I don't object to maintainer edits and I believe they will serve the purpose of smooth cooperation. After the years I'm still unable to get all the coding details right and related pull request conversations tend to be lengthy. Feel free to implement small adjustments on your own. I'd appreciate if the changes were mentioned in the commit message like in [this example](https://kernel.suse.com/cgit/kernel-livepatch/commit/?h=SLE15_Update_15&id=f842fd53a604965a864d3827ab90c691b44cb9a6) (note the remarks in square brackets) but it is not an absolute requirement from my side.

I try to learn from the feedback.
However, we will always need some signaling for handing over, to avoid duplicate work, and to not block other work in the same units.

lpechacek commented:
`Util::pointsFormCorner()` test added. I used the data driven test format. Let me know if that is fine with you or you have a different idea about the test procedure. Please note that I haven't yet checked the test coverage.
I'll respond to your comments later throughout the day.

dg0yt commented:
> `Util::pointsFormCorner()` test added. I used the data driven test format. Let me know if that is fine with you or you have a different idea about the test procedure.

:+1: Looks very good to me.

lpechacek commented:
Changes in this iteration:
* Drop unnecessary `const` from function declarations.
* Reorder `Util::pointsFormCorner()` parameters.
* Add `Util::pointsFormCorner()` test.
